### PR TITLE
Check ambiguous methods in stdlib

### LINKED ISSUE #27408: method ambiguity checks for stdlib packages
The existing method ambiguity check does not cover `stdlib` packages. Consequently method ambiguities have accumulated in some `stdlib` packages (e.g. `LinearAlgebra`, Ref. https://github.com/JuliaLang/julia/pull/27405#issuecomment-394197473). We should check for and fix any method ambiguities and then set up ambiguity checks for all `stdlib` packages to prevent continued regression. Best!

### PR DESCRIPTION
**Pull request by tkf**
closes #27408

---

@Sacha0 I added the tests I suggested in #27408.

The nice thing I noticed is that the number of ambiguous methods is now dropped to 13 (from 25)!

### CODE REVIEW

**Comment by tkf** [comment]
I'm puzzled that my change invokes test failures in very different places:

```
Error in testset ambiguous:
Test Failed at /tmp/julia/share/julia/test/ambiguous.jl:238
  Expression: length(detect_ambiguities(Ambig7)) == 1
   Evaluated: 0 == 1
Error in testset ambiguous:
Test Failed at /tmp/julia/share/julia/test/ambiguous.jl:280
  Expression: length(detect_ambiguities(Ambig9, ambiguous_bottom=true)) == 1
   Evaluated: 0 == 1
ERROR: LoadError: Test run finished with errors
in expression starting at /tmp/julia/share/julia/test/runtests.jl:61
```

--- https://travis-ci.org/JuliaLang/julia/jobs/416266235#L2059


Those happen at the code way after the code I changed.  Also, I don't see how those tests can interact with the one I added:

https://github.com/JuliaLang/julia/blob/d62db6afe88418acdb45d1691d5188256de05dc5/test/ambiguous.jl#L233-L238

https://github.com/JuliaLang/julia/blob/d62db6afe88418acdb45d1691d5188256de05dc5/test/ambiguous.jl#L272-L281

I'll try to debug it but please let me know if I'm making some obvious mistakes.

**Comment by tkf** [comment]
OK.  So tests are all fixed now and CIs are all green.  @Sacha0 Can you merge this?

FYI: The puzzling behavior was due to this silly mistake:
https://github.com/JuliaLang/julia/pull/28665/commits/1721f746a0a9caa0d020f954cb74b280ce3c043a

**Comment by tkf** [comment]
Now that #28749 is merged (thanks @andreasnoack!), there are only two ambiguous methods left (occurring in SparseArrays).

```julia
julia> detect_ambiguities(Base64, CRC32c, Dates, DelimitedFiles, Distributed, FileWatching,
                  Future, InteractiveUtils, Libdl, LibGit2, LinearAlgebra, Logging,
                  Markdown, Mmap, Pkg, Printf, Profile, Random, REPL,
                  Serialization, SHA, SharedArrays, Sockets, SparseArrays, Statistics,
                  SuiteSparse, Test, Unicode, UUIDs; imported=true, recursive=true)
Skipping Distributed.cluster_manager
Skipping Random.DSFMT.win32_SystemFunction036!
2-element Array{Tuple{Method,Method},1}:
 (capturescalars(f, mixedargs::Tuple{Ref{Type{T}},Vararg{Any,N} where N}) where T in SparseArrays.HigherOrderFns at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/SparseArrays/src/higherorderfns.jl:1025, capturescalars(f, mixedargs::Tuple{Union{AbstractArray{0,N} where N, Ref},Ref{Type{T}},Vararg{Any,N} where N}) where T in SparseArrays.HigherOrderFns at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/SparseArrays/src/higherorderfns.jl:1029)
 (_copy(f, args::SparseVector...) in SparseArrays.HigherOrderFns at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/SparseArrays/src/higherorderfns.jl:968, _copy(f, args::SparseMatrixCSC...) in SparseArrays.HigherOrderFns at /home/takafumi/repos/watch/julia/usr/share/julia/stdlib/v1.1/SparseArrays/src/higherorderfns.jl:969)
```

I changed the bound and then rebase onto master.  I may have a go at SparseArrays sometime later but I think we still need this global test to avoid further ambiguities.  If it's solved in SparseArrays we just need to swap `@test_broken` with `@test`.

File: test/ambiguous.jl [hunk 1]
```diff
@@ -157,6 +157,27 @@ ambs = detect_ambiguities(Ambig5)
...
+import InteractiveUtils
+push!(modules, InteractiveUtils)
+
+# At the moment, there are some method ambiguities in stdlibs.  We
+# impose a bound to the number of detected ambiguities to avoid
+# introducing more ambiguities.
+# https://github.com/JuliaLang/julia/issues/27408
+@test length(detect_ambiguities(modules...; imported=true, recursive=true)) ≤ 2
+
+@test_broken detect_ambiguities(modules...; imported=true, recursive=true) == []
```
Comment (Sacha0) [inline test/ambiguous.jl]: Perhaps `isempty`?

**Review by Sacha0 — APPROVED** [review]
LGTM per a brief review! Thanks @tkf! :)

File: test/ambiguous.jl (hunk 1 above)
Comment (tkf) [reply test/ambiguous.jl]: I followed the example for `Base` and `Core` as I thought it makes sense:

https://github.com/JuliaLang/julia/blob/f31e28a50e029942d5d96977f5993defb2f60c13/test/ambiguous.jl#L154-L156

Shall I add a comment to clarify it?

Comment (Sacha0) [reply test/ambiguous.jl]: No need for a comment; it's a matter of style really :). I'm a fan of using evocatively named predicate functions where they exist. Best!

File: test/ambiguous.jl [hunk 2]
```diff
@@ -157,6 +157,28 @@ ambs = detect_ambiguities(Ambig5)
...
+push!(modules, InteractiveUtils)
+
+# At the moment, there are some method ambiguities in stdlibs.  We
+# impose a bound to the number of detected ambiguities to avoid
+# introducing more ambiguities.
+# https://github.com/JuliaLang/julia/issues/27408
+@test length(detect_ambiguities(modules...; imported=true, recursive=true)) ≤ 2
+
+# not using isempty so this prints more information when it fails
+@test_broken detect_ambiguities(modules...; imported=true, recursive=true) == []
```
[the remark this answers is not in the document]
Comment (Sacha0) [reply test/ambiguous.jl]: Oh gosh, apologies, I completely missed the relevant comment in the code snippet you linked above! Absolutely, good call.

[the remark this answers is not in the document]
Comment (tkf) [reply test/ambiguous.jl]: The relevant code for `Base` and `Core` is something like 20 lines above so I think it makes sense to repeat the comment.  Someone replacing `@test_broken` with `@test` might just swap it with `isempty` if there is no comment.

[the remark this answers is not in the document]
Comment (Sacha0) [reply test/ambiguous.jl]: Agreed! :)

**Comment by tkf** [comment]
Now that #30120 is merged, there is no method ambiguities in stdlib. Isn't it the best timing to merge this?

File: test/ambiguous.jl [hunk 3]
```diff
@@ -157,6 +157,22 @@ ambs = detect_ambiguities(Ambig5)
...
 @test !isempty(detect_ambiguities(Core, Base; imported=true, ambiguous_bottom=true))
 
+module AmbigStdlib
+using Test
+
+for lib in Base._stdlibs
+    @eval import $lib
+end
+
+modules = @eval [$(Base._stdlibs...)]
```
[the remark this answers is not in the document]
Comment (tkf) [reply test/ambiguous.jl]: Thanks. Good point. It makes the patch simpler.

**Comment by chriscoey** [comment]
bump

**Comment by vtjnash** [comment]
Whoops, yes we should have. It needs to be rebased though now, so we can check if it is still working.

**Comment by tkf** [comment]
There is no conflicts so I guess re-running CI is sufficient?

**Comment by vtjnash** [comment]
It needs to be rebased to get a new merge commit

**Comment by tkf** [comment]
Squashed & rebased.

**Pull request merged**